misc/python/materialize/zippy/source_actions.py

```diff
@@ -86,3 +109,32 @@ def run(self, c: Composition) -> None:
 
     def provides(self) -> list[Capability]:
         return [self.source]
+
+
+class AlterSourceConnection(Action):
+    def __init__(self, capabilities: Capabilities, source: SourceExists) -> None:
+        self.source = source
+        super().__init__(capabilities)
+
+    def run(self, c: Composition) -> None:
+        # This flips the usage of the SSH tunnel.
+        self.flip_usage_of_ssh_tunnel(
+            c, new_use_ssh_status=not self.source.uses_ssh_tunnel
+        )
+        self.source.uses_ssh_tunnel = not self.source.uses_ssh_tunnel
+
+    def flip_usage_of_ssh_tunnel(
+        self, c: Composition, new_use_ssh_status: bool
+    ) -> None:
+        kafka_connection_name = f"{self.source.name}_kafka_conn"
+        c.testdrive(
+            dedent(
+                f"""
+                > ALTER CONNECTION {kafka_connection_name} SET BROKER '${{testdrive.kafka-addr}}'
+                  {'USING SSH TUNNEL zippy_ssh' if new_use_ssh_status else ''};
+                """
+            )
+        )
+
+    def provides(self) -> list[Capability]:
```

Comment:
AlterSourceConnection does not provide any additional capabilities, it merely operates on existing objects. So `provides()` needs to return an empty set (or not be defined at all, so the superclass provides() applies, which returns an empty set. This explains the "source specified more than once" error you are getting in the CI.